superset-websocket/src/index.ts

```diff
@@ -0,0 +1,329 @@
+
+interface ChannelValue {
+  sockets: Array<string>,
+}
+
+const environment = process.env.NODE_ENV;
+export const opts = {
+  port: 8080,
+  redis: {
+    port: 6379,
+    host: "127.0.0.1",
+    password: "",
+    db: 0
+  },
+  streamPrefix: "async-events-",
+  jwtSecret: "",
+  jwtCookieName: "async-token",
+  redisStreamReadCount: 100,
+  redisStreamReadBlockMs: 5000,
+  socketResponseTimeoutMs: 60 * 1000,
+  pingSocketsIntervalMs: 20 * 1000,
+  gcChannelsIntervalMs: 120 * 1000,
+}
+
+const startServer = process.argv[2] === 'start';
+const configFile = environment === 'test' ? '../config.test.json' : '../config.json';
+let config = {};
+try {
+  config = require(configFile);
+} catch(err) {
+  console.warn('config.json not found, using defaults');
+}
+
+Object.assign(opts, config);
+
+if(startServer && opts.jwtSecret.length < 32)
+  throw('Please provide a JWT secret at least 32 bytes long')
+
+const redis = new Redis(opts.redis);
+const httpServer = http.createServer();
+export const wss = new WebSocket.Server({ noServer: true, clientTracking: false });
+
+const SOCKET_ACTIVE_STATES = [WebSocket.OPEN, WebSocket.CONNECTING];
+const GLOBAL_EVENT_STREAM_NAME = `${opts.streamPrefix}full`;
+const DEFAULT_STREAM_LAST_ID = '$';
+
+export let channels: Record<string, ChannelValue> = {};
+export let sockets: Record<string, SocketInstance> = {};
+let lastFirehoseId: string = DEFAULT_STREAM_LAST_ID;
+
+
+export const setLastFirehoseId = (id: string): void => {
+  lastFirehoseId = id;
+}
+
+export const trackClient = (channel: string, socketInstance: SocketInstance): string => {
+  const socketId = uuidv4();
+  sockets[socketId] = socketInstance;
+
+  if(channel in channels) {
+    channels[channel].sockets.push(socketId)
+  } else {
+    channels[channel] = {sockets: [socketId]};
+  }
+
+  return socketId;
+}
+
+export const sendToChannel = (channel: string, value: EventValue): void => {
+  const strData = JSON.stringify(value);
+  if(!channels[channel]) {
+    console.debug(`channel ${channel} is unknown, skipping`);
+    return;
+  }
+  channels[channel].sockets.forEach(socketId => {
+    const socketInstance: SocketInstance = sockets[socketId];
+    if(!socketInstance) return cleanChannel(channel);
+    try {
+      socketInstance.ws.send(strData);
+    } catch(err) {
+      console.debug('Error sending to socket', err);
+      cleanChannel(channel);
+    }
+  });
+}
+
+export const fetchRangeFromStream = async ({sessionId, startId, endId, listener}: FetchRangeFromStreamParams) => {
+  const streamName = `${opts.streamPrefix}${sessionId}`;
+  try {
+    const reply = await redis.xrange(streamName, startId, endId);
+    if (!reply || !reply.length) return;
+    listener(reply);
+  } catch(e) {
+    console.error(e);
+  }
+}
+
+export const subscribeToGlobalStream = async (stream: string, listener: ListenerFunction) => {
+  /*eslint no-constant-condition: ["error", { "checkLoops": false }]*/
+  while (true) {
```

Comment:
It's probably premature to worry about performance here and I think this is likely not an issue, but I'm curious if there are any potential concerns you foresee with an infinite loop calling redis? It looks like the BLOCK option should prevent too many requests in too short a time since if no events are available it'll wait, but not sure what would happen if one or two events regularly trickle in within short timeframes?

Reply:
When events are received, it returns immediately. More info here: https://redis.io/commands/xread#blocking-for-data
> When the BLOCK command is passed, but there is data to return at least in one of the streams passed, the command is executed synchronously exactly like if the BLOCK option would be missing.